src/Resources/contao/controllers/BackendMain.php

```diff
@@ -293,4 +279,48 @@ protected function output()
 
 		return $this->Template->getResponse();
 	}
+
+	/**
+	 * Adjusts the logout link if the current user is impersonated.
+	 */
+	private function setImporsonatedLogout()
+	{
+		$token = \System::getContainer()->get('security.token_storage')->getToken();
+
+		if (!$token instanceof TokenInterface)
+		{
+			return;
+		}
+
+		$impersonatorUser = null;
+		foreach ($token->getRoles() as $role)
+		{
+			if ($role instanceof SwitchUserRole)
+			{
+				$impersonatorUser = $role->getSource()->getUsername();
+				break;
+			}
+		}
+
+		if (!$impersonatorUser)
+		{
+			return;
+		}
+
+		$firewallMap = \System::getContainer()->get('security.firewall.map');
+		$request = \System::getContainer()->get('request_stack')->getCurrentRequest();
+
+		// generate exit impersonation path from current request
+		if (null === $request || null === ($firewallConfig = $firewallMap->getFirewallConfig($request)) || null === ($switchUserConfig = $firewallConfig->getSwitchUser()))
```

Comment:
Again, the null pointer handling is wrong. We expect that there is always a request, so we should throw an exception if there is not.

Reply:
Why would we throw an exception if the application works perfectly fine without it? It would just generate the regular logout link …

Reply:
How can there ever not be a request?

Reply:
I don't think that case will ever happen. It's just shorter code 😄